PKG-5550: Add gguf package & Update to 0.9.1

jnoller opened the pull request:
gguf 0.9.1

**Destination channel:** ai-staging

### Links

- [PKG-5550](https://anaconda.atlassian.net/browse/PKG-5550) 
- [Upstream repository](https://github.com/ggerganov/llama.cpp/tree/master/gguf-py)
- [Upstream changelog/diff](https://pypi.org/project/gguf/#history)

### Explanation of changes:

- Updates the forked gguf-feedstock recipe to match the latest published PyPI package: https://pypi.org/project/gguf/ (0.9.1)
- Pins dependency versions, adds missing `sentencepiece` dependency


[PKG-5550]: https://anaconda.atlassian.net/browse/PKG-5550?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

jnoller committed:
Update to 0.9.1, pin deps, fix entrypoints

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(125): `conda run conda-lint /tmp/abs_e5fmpk5tml/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:49: missing_description: The recipe is missing a description
- clone/recipe/meta.yaml:20: avoid_noarch: noarch: python packages should be avoided

===== ERRORS =====
- clone/recipe/meta.yaml:49: missing_documentation: The recipe is missing a doc_url or doc_source_url
- clone/recipe/meta.yaml:49: missing_license_family: The recipe is missing the `about/license_family` key.
- clone/recipe/meta.yaml:49: missing_dev_url: The recipe is missing a dev_url
===== Final Report: =====
3 Errors and 2 Warnings were found

</details>

JeanChristopheMorinPerso commented:
@jnoller You will need to create a file named `abs.yaml` at the root of the repo with this content in order to upload the packages to `ai-staging`:
```yaml
upload_channels:
- ai-staging
```
You can add it into this PR.

jnoller committed:
Add abs.yaml

jnoller commented:
@JeanChristopheMorinPerso Added abs.yaml to the PR

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(125): `conda run conda-lint /tmp/abs_0cr6exo34o/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:20: avoid_noarch: noarch: python packages should be avoided
- clone/recipe/meta.yaml:49: missing_description: The recipe is missing a description

===== ERRORS =====
- clone/recipe/meta.yaml:49: missing_documentation: The recipe is missing a doc_url or doc_source_url
- clone/recipe/meta.yaml:49: missing_dev_url: The recipe is missing a dev_url
- clone/recipe/meta.yaml:49: missing_license_family: The recipe is missing the `about/license_family` key.
===== Final Report: =====
3 Errors and 2 Warnings were found

</details>

JeanChristopheMorinPerso commented on recipe/meta.yaml:
```diff
@@ -1,40 +1,47 @@
+    - gguf-dump = scripts:gguf_dump_entrypoint
+    - gguf-set-metadata = scripts:gguf_set_metadata_entrypoint
+    - gguf-new-metadata = scripts:gguf_new_metadata_entrypoint
   noarch: python
```
```suggestion
  skip: true  # [py<38]
```
We generally don't build noarch-python packages.

jnoller replied:
I just committed this exact change 👍

JeanChristopheMorinPerso commented on recipe/meta.yaml:
```diff
@@ -1,40 +1,47 @@
 
 requirements:
   host:
     - python >=3.8
```
```suggestion
    - python
```

JeanChristopheMorinPerso commented on recipe/meta.yaml:
```diff
@@ -1,40 +1,47 @@
     - pip
-    - poetry-core
   run:
     - python >=3.8
```
```suggestion
    - python
```

jnoller committed:
Fix linter errors, add description, dev and other urls

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(125): `conda run conda-lint /tmp/abs_ffpap9l4ao/clone` failed. (See above for error)
The following problems have been found:

===== WARNINGS =====
- clone/recipe/meta.yaml:20: avoid_noarch: noarch: python packages should be avoided
===== Final Report: =====
0 Errors and 1 Warning were found

</details>

JeanChristopheMorinPerso commented on recipe/meta.yaml:
```diff
@@ -1,46 +1,59 @@
   license: MIT
+  license_family: MIT
   license_file: LICENSE
+  dev_url: https://github.com/ggerganov/llama.cpp
```
```suggestion
  dev_url: https://github.com/ggerganov/llama.cpp/tree/master/gguf-py
```
When a repo contains multiple different things, we try to point to the exact folder that contains the package.

JeanChristopheMorinPerso commented on recipe/meta.yaml:
```diff
@@ -1,46 +1,59 @@
   requires:
     - pip
-
+  
```
```suggestion
```

JeanChristopheMorinPerso commented on recipe/meta.yaml:
```diff
@@ -1,46 +1,59 @@
     - python >=3.8
     - numpy >=1.17
+    - tqdm >=4.27
+    - sentencepiece >=0.2.0
```
Can you add a comment to mention that upstream is missing this dependency? I found it in https://github.com/ggerganov/llama.cpp/blob/gguf-v0.9.1/gguf-py/gguf/vocab.py#L10. You night as well create an issue upstream to tell them they are missing a dep in their package metadata.

jnoller replied:
@JeanChristopheMorinPerso Good call out - it's actually defined really weirdly in upstream - the tell users to install all the requirements here: https://github.com/ggerganov/llama.cpp/tree/master/requirements but for the gguf-py pyproject.toml it's missing. Will note and file upstream

jnoller replied:
@JeanChristopheMorinPerso Added, opened a PR in upstream here: https://github.com/ggerganov/llama.cpp/pull/9107

jnoller committed:
Remove noarch, set to skip < py38

jnoller committed:
Update recipe/meta.yaml

Co-authored-by: Jean-Christophe Morin <38703886+JeanChristopheMorinPerso@users.noreply.github.com>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
ERROR conda.cli.main_run:execute(125): `conda run conda-lint /tmp/abs_e380gbzfaw/clone` failed. (See above for error)
The following problems have been found:

===== ERRORS =====
- clone/recipe/meta.yaml:31: remove_python_pinning: On arch specific packages, python deps should not be constrained.
===== Final Report: =====
1 Error and 0 Warnings were found

</details>

jnoller committed:
Update recipe/meta.yaml

Co-authored-by: Jean-Christophe Morin <38703886+JeanChristopheMorinPerso@users.noreply.github.com>

jnoller committed:
Update recipe/meta.yaml

Co-authored-by: Jean-Christophe Morin <38703886+JeanChristopheMorinPerso@users.noreply.github.com>

jnoller committed:
Update recipe/meta.yaml

Co-authored-by: Jean-Christophe Morin <38703886+JeanChristopheMorinPerso@users.noreply.github.com>

jnoller committed:
Add note for upstream PR / issue

ryanskeith commented on recipe/meta.yaml:
```diff
@@ -1,46 +1,59 @@
+    - python
     - numpy >=1.17
+    - tqdm >=4.27
+    - sentencepiece >=0.2.0 # This dependency is missing in the upstream pyproject.toml file: https://github.com/ggerganov/llama.cpp/blob/master/gguf-py/pyproject.toml
```
```suggestion
    - pyyaml >=5.1
    - sentencepiece >=0.2.0 # This dependency is missing in the upstream pyproject.toml file: https://github.com/ggerganov/llama.cpp/blob/master/gguf-py/pyproject.toml
```

According to this line: https://github.com/ggerganov/llama.cpp/blob/2f3c1466ff46a2413b0e363a5005c46538186ee6/gguf-py/pyproject.toml#L25

Jrice1317 replied:
It would seem the correct `pyyproject.toml` is [this](https://github.com/ggerganov/llama.cpp/blob/gguf-v0.9.1/gguf-py/pyproject.toml) one and there doesn't seem to be a mention of `pyyaml`.

jnoller replied:
Thank you @ryanskeith and @Jrice1317 - I just caught that too, they're missing pyyaml but not sure how my build passed missing it. I also just realized there's an issue linking / using the pypi package version vs the upstream source. 

I'm going to commit a fix that moves to the upstream source (and add pyyaml) instead of the published pypi package and use a version scheme similar to llama.cpp-feedstock.

JeanChristopheMorinPerso replied:
What the problem with the version @jnoller? https://inspector.pypi.io/project/gguf/0.9.1/packages/9f/1e/6bd024e0138d663cd333e8fde4c03d343e8be680d43b86537ef1497a7e32/gguf-0.9.1.tar.gz/gguf-0.9.1/pyproject.toml matches  https://github.com/ggerganov/llama.cpp/blob/gguf-v0.9.1/gguf-py/pyproject.toml from what I see...

ryanskeith reviewed (changes requested):
Please add the missing dependency (unless I am looking in the wrong place).  Other than that looks good.

jnoller committed:
Move to source repo and match upstream build version, add pyyaml dep

jnoller commented:
@JeanChristopheMorinPerso @Jrice1317 @ryanskeith please re-review with the latest changes. I added pyyaml, but also moved to using the llama.cpp project's releases directly - the pypi project is pretty out of date and to support adding the llama-cpp-scripts package it makes sense to pin all three (include llama.cpp-feedstock) the same way. The llama-cpp-scripts pyproject.toml in the root of the llama.cpp tree has always had a version of 0.0.0 but it has a dependency on a matching version of the gguf package. This way all 3 packages work from the source repo directly for now.

jnoller commented:
I'm closing this PR: After working on packaging the llama-cpp-scripts work and working through the dependency tree, I think it makes more sense collapse the `gguf` bundling/packaging within the `llama-cpp-scripts` packaging as the two are tightly coupled. It's better to ignore the upstream pyproject.toml files and follow the instructions to set up quantize and other tools (https://github.com/ggerganov/llama.cpp/blob/master/examples/quantize/README.md). Its unclear if the pyproject.toml files are the recommended path forward (https://github.com/ggerganov/llama.cpp/pull/9107).
